### PR TITLE
mingw cross-compilation issues

### PR DESCRIPTION
**Pull request by aitap**
Two issues arise when cross-compiling dmalloc with `--build=i686-linux-gnu --host=i686-w64-mingw32`:

 * `AC_RUN_IFELSE` tests for `getpagesize()` don't have a third fallback argument, stopping `./configure` with a fatal error
 * `Makefile.in` adds `.` to the list of system include directories (i.e. passes `-I.`), which results in local `./malloc.h` being included in a system header when it requests `#include <malloc.h>`

The former problem makes dmalloc impossible to cross-compile; the latter seems to only occur on mingw, but could also (in theory) occur with a different libc which has `#include <malloc.h>` in its public headers.

These commits fix both problems.

### CODE REVIEW

**Comment by bharath999** [comment]
I have patched this change (**AC_RUN_IFELSE tests for getpagesize() don't have a third fallback argument, stopping ./configure with a fatal error**)
I am still facing error while running the configure file. I am compiling the code for ARM processor.
I have attached the configure error in the text file

[dmalloc_configure_error.txt](https://github.com/j256/dmalloc/files/5701234/dmalloc_configure_error.txt)

Please can you help me how to overcome this "dmalloc" configuration issue.
Thanks in advance.

**Comment by aitap** [comment]
I have patched only configure.ac, not the configure script itself.
Please run autoconf to regenerate the configure script from
configure.ac.

-- 
Best regards,
Ivan

**Comment by j256** [comment]
Yeah I can't take this with the Makefile.in changes.  Typically you are supposed to be able to load the sources into one directory and build into another.  If I do the following, the build breaks without the `-I$(srcdir)`.

```
make distclean
mkdir tmp
cd tmp
../configure
make
```

**Comment by j256** [comment]
When is the system's malloc.h included?  Is it brought it through configure somehow?

File: Makefile.in [hunk 1]
```diff
@@ -31,7 +31,7 @@ DEFS = -DHAVE_STDARG_H=@HAVE_STDARG_H@ \
 	-DHAVE_SYS_CYGWIN_H=@HAVE_SYS_CYGWIN_H@ \
 	-DHAVE_SIGNAL_H=@HAVE_SIGNAL_H@ \
 	$(DEFINES)
-INCS = -I. -I$(srcdir) $(INCLUDES)
```
Comment (j256) [inline Makefile.in:34]: We may be able to do without the `-I$(srcdir)` but we need the `-I.` to be able to find the local headers.

**Comment by j256** [comment]
Wait.  I don't understand the getpagesize fixes.   configure.ac sets `ac_cv_page_size` to 0 at the top and then only updates if one of the test programs works.  What is the error?

Also, please submit multiple issues/pulls.

**Comment by j256** [comment]
Fixed the getpagesize stuff on master.  See: https://github.com/j256/dmalloc/pull/30

**Comment by aitap** [comment]
Thanks for fixing the getpagesize problems! Regarding system `malloc.h`, here's what I see when cross-compiling with `-I.`:

```
x86_64-w64-mingw32-gcc -g -O2  -DHAVE_STDARG_H=1 -DHAVE_STDLIB_H=1 -DHAVE_STRING_H=1 -DHAVE_UNISTD_H=1 -DHAVE_SYS_MMAN_H=0 -DHAVE_SYS_TYPES_H=1 -DHAVE_W32API_WINBASE_H=0 -DHAVE_W32API_WINDEF_H=0 -DHAVE_SYS_CYGWIN_H=0 -DHAVE_SIGNAL_H=1  -I. -I.  -c arg_check.c -o ./arg_check.o
In file included from /usr/share/mingw-w64/include/stdlib.h:695:0,
                 from arg_check.c:34:
./malloc.h:90:1: error: unknown type name ‘DMALLOC_PNT’
 DMALLOC_PNT dmalloc_malloc(const char *file, const int line,
 ^~~~~~~~~~~
```

mingw's `stdlib.h` line 695 is `#include <malloc.h>`.

**Comment by aitap** [comment]
Hmm, `malloc.h` resides in `$(srcdir)`, not `.`, and it's not even `#include`d anywhere. Moreover, the *.c files already use `#include "..."` for files residing in `$(srcdir)`, so there is no need for `-I$(srcdir)` (unlike `-I.`).

It should be enough to either
 * remove `-I$(srcdir)`
 * rename `malloc.h` it to `malloc.h.src` or something equally unlikely to match a system header name

I see that the following patch doesn't break the subdirectory build for me, either native or cross-compiled with MinGW, i.e. `mkdir -p temp && cd temp && ../configure && make` succeeds with or without `--host=x86_64-w64-mingw32`:
```diff
diff --git a/Makefile.in b/Makefile.in
index 926048a..c87edef 100644
--- a/Makefile.in
+++ b/Makefile.in
@@ -31,7 +31,7 @@ DEFS = -DHAVE_STDARG_H=@HAVE_STDARG_H@ \
        -DHAVE_SYS_CYGWIN_H=@HAVE_SYS_CYGWIN_H@ \
        -DHAVE_SIGNAL_H=@HAVE_SIGNAL_H@ \
        $(DEFINES)
-INCS = -I. -I$(srcdir) $(INCLUDES)
+INCS = -I. $(INCLUDES)
 LIBS = -L. @LIBS@
 A_OUT = aout@EXEEXT@
 INFOFILE = dmalloc.info
```

**Comment by aitap** [comment]
Sorry, removing only `-I$(srcdir)` will still break the build when `$(srcdir)` = `.`. I'm afraid the only viable solution left is renaming malloc.h and patching the `dmalloc.h.3 :` target in `Makefile.in`.

**Comment by j256** [comment]
Yeah I just did that.  See: https://github.com/j256/dmalloc/pull/32

**Comment by aitap** [comment]
Yes, thanks a lot!

**Pull request closed without merging**